File: programmers/다리를 지나는 트럭.py
"""
코딩테스트 연습 > 스택/큐
"""

def solution(bridge_length, weight, truck_weights):
    bridge = [None for _ in range(bridge_length)]

    bridge_weight_sum = 0
    count = 0
    while True:
        count += 1
        last = bridge.pop()
        if last:
            bridge_weight_sum -= last

        if len(truck_weights) != 0:
            target_truck = truck_weights[0]

            if target_truck <= weight - bridge_weight_sum:
                truck_weights = truck_weights[1:]
                bridge_weight_sum += target_truck
                bridge = [target_truck] + bridge
            else:
                bridge = [None] + bridge

        if bridge_weight_sum == 0:
            return count

File: programmers/다리를 지나는 트럭.kt
import java.util.*

class Solution {
    fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
        var bridge = ArrayDeque<Int>()
        var truck_weight_sum = 0
        var trucks = ArrayDeque<Int>()

        trucks.addAll(truck_weights.toList())

        bridge.addAll(
            (0..bridge_length-1).map{ 0 }
        )

        var time = 0
        while (true) {
            if (truck_weight_sum == 0 && trucks.size == 0) {
                break
            }
            time += 1

            var end_truck = bridge.pollLast()
            truck_weight_sum = truck_weight_sum - end_truck

            var cand: Int? = trucks.peekFirst()
            if (cand != null && truck_weight_sum + cand <= weight) {
                trucks.pollFirst()
                truck_weight_sum += cand
                bridge.addFirst(cand)
            } else {
                bridge.addFirst(0)
            }
        }

        return time
    }
}
